src/aspire/abinitio/commonline_cn.py

```diff
@@ -0,0 +1,367 @@
+            c2s_inds[inds] += n_theta // 2
+            c2s_inds[inds] = np.mod(c2s_inds[inds], n_theta)
+
+            scls_inds[i_cand, :, 0] = c1s_inds
+            scls_inds[i_cand, :, 1] = c2s_inds
+        return scls_inds
+
+    # TODO: cache
+    def _compute_cls_inds(self, Ris_tilde, R_theta_ijs):
+        """
+        Compute the common-lines indices induced by the candidate rotations.
+
+        :param Ris_tilde: An array of size n_candsx3x3 of candidate rotations.
+        :param R_theta_ijs: An array of size n_theta_ijsx3x3 of inplane rotations.
+        :return: An array of size n_cands x n_cands x n_theta_ijs x 2 holding common-lines
+            indices induced by the supplied rotations.
+        """
+        n_theta = self.n_theta
+        n_points_sphere = self.n_points_sphere
+        n_theta_ijs = R_theta_ijs.shape[0]
+        cij_inds = np.zeros(
+            (n_points_sphere, n_points_sphere, n_theta_ijs, 2), dtype=np.uint16
+        )
+        logger.info("Computing common-line indices induced by candidate rotations.")
+        with tqdm(total=n_points_sphere) as pbar:
+            for i in range(n_points_sphere):
+                for j in range(n_points_sphere):
+                    R_cands = Ris_tilde[i].T @ R_theta_ijs @ Ris_tilde[j]
+
+                    c1s = np.array((-R_cands[:, 1, 2], R_cands[:, 0, 2])).T
+                    c2s = np.array((R_cands[:, 2, 1], -R_cands[:, 2, 0])).T
+
+                    c1s = self.cl_angles_to_ind(c1s, n_theta)
+                    c2s = self.cl_angles_to_ind(c2s, n_theta)
+
+                    inds = np.where(c1s >= n_theta // 2)
+                    c1s[inds] -= n_theta // 2
+                    c2s[inds] += n_theta // 2
+                    c2s[inds] = np.mod(c2s[inds], n_theta)
+
+                    cij_inds[i, j, :, 0] = c1s
+                    cij_inds[i, j, :, 1] = c2s
+                pbar.update()
+        return cij_inds
+
+    def _generate_cand_rots(self):
+        logger.info("Generating candidate rotations.")
+        # Construct candidate rotations, Ris_tilde.
+        vis = self._generate_cand_rots_third_rows()
+        Ris_tilde = np.array([self._complete_third_row_to_rot(vi) for vi in vis])
+
+        # Construct all in-plane rotations, R_theta_ijs
+        # The number of R_theta_ijs must be divisible by the symmetric order.
+        n_theta_ij = 360 - (360 % self.order)
+        theta_ij = np.arange(0, n_theta_ij, self.degree_res) * np.pi / 180
+        R_theta_ijs = Rotation.about_axis("z", theta_ij).matrices
+
+        return Ris_tilde, R_theta_ijs
+
+    def _generate_cand_rots_third_rows(self, legacy=True):
+        n_points_sphere = self.n_points_sphere
+        if legacy:
+            # Genereate random points on the sphere
+            third_rows = randn(n_points_sphere, 3, seed=self.seed)
+            third_rows /= anorm(third_rows, axes=(-1,))[:, np.newaxis]
+        else:
+            # Use Fibonocci sphere points
```

Comment:
Ended up refactoring `_generate_cand_rots` to build rotations on the fly while excluding rotations that represent equator images. This we generate `self.n_points_sphere` usable rotations instead of discarding them after creation.

I removed the Fibonocci stuff and saved it for later. It performed similarly to the randomly chosen third rows.